app/controllers/api/v1/bookmarks_controller.rb

```diff
@@ -0,0 +1,127 @@
+class Api::V1::BookmarksController < Api::V1::BaseController
+  respond_to :json
+
+  def create
+    archivist = User.find_by_login(params[:archivist])
+    bookmarks = params[:bookmarks]
+    bookmarks_responses = []
+    @bookmarks = []
+
+    # check for top-level errors (not an archivist, no bookmarks...)
+    status, messages = batch_errors(archivist, bookmarks)
+
+    if status == :ok
+      # Flag error and successes
+      @some_errors = @some_success = false
+
+      # Process the works, updating the flags
+      bookmarks.each do |bookmark|
+        bookmarks_responses << import_bookmark(archivist, bookmark)
+      end
+
+      # set final response code and message depending on the flags
+      messages = response_message(messages)
+    end
+
+    render status: status, json: { status: status, messages: messages, bookmarks: bookmarks_responses }
+  end
+
+  private
+
+  # Set messages based on success and error flags
+  def response_message(messages)
+    if @some_success && @some_errors
+      messages << "At least one bookmark was not created. Please check the individual bookmark results for further information."
+    elsif !@some_success && @some_errors
+      messages << "None of the bookmarks were created. Please check the individual bookmark results for further information."
+    else
+      messages << "All bookmarks were successfully created."
+    end
+    messages
+  end
+
+  # Returns a hash
+  def import_bookmark(archivist, params)
+    bookmark_request = external_bookmark(archivist, params)
```

Comment:
Sorry to be nitpicky some more: it feels to me like even here it makes more sense to have this remain generic (eg "build_bookmark_request(user, params)") because the only place where it actually becomes specific to external works is in the currently named external_bookmark method and bookmark_errors (which is already named generically) -- and in those two places, I would suggest adding a comment explaining that for the moment this ONLY works for bookmarks for external works. There is basically no difference except if the "external" stuff is passed AFAIK.